### PR TITLE
Dynamic exporter buckets

### PR DESCRIPTION
**Pull request by skaes**
WIP

### CODE REVIEW

File: go/prometheusexporter/collector/collector.go [hunk 1]
```diff
@@ -171,6 +212,7 @@ func (c *Collector) observeMetrics(m *metric) {
 func (c *Collector) observer() {
 	for !util.Interrupted() && atomic.LoadUint32(&c.stopped) == 0 {
 		m := <-c.metricsChannel
+		c.mutex.Lock()
```
Comment (0robustus1) [inline go/prometheusexporter/collector/collector.go]: Shouldn't there be a `defer c.mutex.Unlock()` after this line? Searching through the entire file the calls between mutex.Lock and mutex.Unlock seem unbalanced (off by one).

File: go/prometheusexporter/collector/collector.go [hunk 2]
```diff
@@ -201,6 +243,18 @@ func labelsFromLabelPairs(pairs []*promclient.LabelPair) prometheus.Labels {
 	return labels
 }
 
+func (c *Collector) deleteSummaryLabels(summary *prometheus.SummaryVec, labels prometheus.Labels) bool {
+	c.mutex.Lock()
```
Comment (0robustus1) [inline go/prometheusexporter/collector/collector.go]: I'm a bit unsure why we need to lock the mutex here. We are not accessing the stream and we are just changing the value of the summary itself. This shouldn't have an impact to the pointer maintained by the `c *Collector` or the registration of said summary in the registry.

File: go/prometheusexporter/collector/collector.go [hunk 3]
```diff
@@ -201,6 +243,18 @@ func labelsFromLabelPairs(pairs []*promclient.LabelPair) prometheus.Labels {
...
+	defer c.mutex.Unlock()
+	return summary.Delete(labels)
+}
+
+func (c *Collector) deleteHistogramLabels(histogram *prometheus.HistogramVec, labels prometheus.Labels) bool {
+	c.mutex.Lock()
```
Comment (0robustus1) [inline go/prometheusexporter/collector/collector.go]: Same case as mentioned for the summary. I don't think that this needs to be guarded by the mutex that was introduced to guard the stream config in the collector.

File: go/prometheusexporter/collector/collector.go (hunk 1 above)
Comment (skaes) [reply go/prometheusexporter/collector/collector.go]: This is a left over from refactoring the code mid change. Thx for spotting!

File: go/prometheusexporter/collector/collector.go (hunk 2 above)
Comment (skaes) [reply go/prometheusexporter/collector/collector.go]: You might be right. In any case, the lock is in the wrong place. The pointer access to the histogramvec needs to be protected, but the code doesn't do this.

File: go/prometheusexporter/collector/collector.go (hunk 3 above)
Comment (skaes) [reply go/prometheusexporter/collector/collector.go]: histogramvecs might be deleted from the collector state in function Update(). That's why we need to protect the pointer access. But that has happened outside the function. Maybe brain must have been half fried when I wrote this.

**Comment by skaes** [comment]
@0robustus1 I implemented the RWMutex suggestion and benchmarked mutex operations. I think we're good to go.

**Pull request merged**